src/MacroExpander.js

```diff
@@ -17,114 +17,194 @@ class MacroExpander {
     }
 
     /**
-     * Recursively expand first token, then return first non-expandable token.
+     * Return the next unexpanded token without removing anything from the
+     * stack.  Similar in behavior to TeX's `\futurelet`.
+     */
+    future() {
+        if (this.stack.length === 0) {
+            this.stack.push(this.lexer.lex());
+        }
+        return this.stack[this.stack.length - 1];
+    }
+
+    /**
+     * Remove and return the next unexpanded token.
+     */
+    popToken() {
+        this.future();  // ensure non-empty stack
+        return this.stack.pop();
+    }
+
+    /**
+     * Consume all following space tokens, without expansion.
+     */
+    consumeSpaces() {
+        for (;;) {
+            const token = this.future();
+            if (token.text === " ") {
+                this.stack.pop();
+            } else {
+                break;
+            }
+        }
+    }
+
+    /**
+     * Expand next token only once, and leave it on the stack.
+     * Returns the token or its expansion.
+     * Used to implement `get` and `expandAfter`.
      *
      * At the moment, macro expansion doesn't handle delimited macros,
      * i.e. things like those defined by \def\foo#1\end{…}.
      * See the TeX book page 202ff. for details on how those should behave.
      */
-    nextToken() {
-        for (;;) {
-            if (this.stack.length === 0) {
-                this.stack.push(this.lexer.lex());
+    expandOnce() {
+        const topToken = this.popToken();
+        const name = topToken.text;
+        const macro = (name.charAt(0) === "\\");
+        if (macro) {
+            // Consume all spaces after \macro
+            this.consumeSpaces();
+        }
+        if (!(macro && this.macros.hasOwnProperty(name))) {
+            // Fully expanded
+            this.stack.push(topToken);
+            return topToken;
```

Comment:
This is a little weird to pop `topToken` at the start just to push it back on.  I can't think of demonstrably better way to do this though.

Reply:
I agree, unfortunately on both counts.  If I didn't have to `consumeSpaces`, I could use `future()` and then `popToken` after the fully expanded case.  Maybe there's another place to put the `consumeSpaces`... but at least this works.